content/articles/creating-a-callback-url-for-safaricom-daraja-api-with-firebase-cloud-functions-in-android/index.md

```diff
@@ -0,0 +1,418 @@
+
+```
+<application>
+...
+        <service
+            android:name=".MessagingService"
+            android:stopWithTask="false"
+            android:exported="false">
+            <intent-filter>
+                <action android:name="com.google.firebase.MESSAGING_EVENT" />
+            </intent-filter>
+        </service>
+</application>
+```
+
+### Step 9 - Write MainActivity Code
+In this step, we will write codes that will make use of the [android-mpesa-api](https://github.com/jumaallan/android-mpesa-api) to integrate payment into our app.
+
+#### Initializing the Daraja API
+Here, replace the SECRET_KEY and the CONSUMER_SECRET_KEY with the ones that you were given when you created an App in the Safaricom Developers portal.
+
+```Kotlin
+        daraja = Daraja.with("CONSUME_KEY", "CONSUMER_SECRET_KEY", Env.SANDBOX,
+            object : DarajaListener<AccessToken> {
+                override fun onResult(result: AccessToken) {
+                    Toast.makeText(applicationContext, result.access_token, Toast.LENGTH_SHORT).show()
+                }
+
+                override fun onError(error: String?) {
+                    Toast.makeText(applicationContext, error.toString(), Toast.LENGTH_SHORT).show()
+                }
+
+            })
+```
+
+When a user clicks the button, we need the following code to be executed:
+```Kotlin
+        findViewById<Button>(R.id.button).setOnClickListener {
+            val phoneNumber = phoneNum.text.toString()
+            val lnmExpress = LNMExpress(
+                "174379",
+                "bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919",
+                TransactionType.CustomerPayBillOnline,
+                "1",
+                phoneNumber,
+                "174379",
+                phoneNumber,
+                "https://us-central1-lnmcallback-c79b4.cloudfunctions.net/api/CallbackUrl",
+                "001ABC",
+                "Goods Payment"
+            )
+            ....
+```
+
+Here, we pass the content of the phone number EditText and also use 1 as the default amount for the transaction.
+
+> Now it is high time to go to your Firebase Console in the Functions section and copy the URL for our API that was generated.
+
+[!Function](section-engineering/creating-a-callback-url-for-safaricom-daraja-api-with-firebase-cloud-functions-in-android/functions_console.png)
+
+In this case, mine is :
+
+`https://us-central1-lnmcallback-c79b4.cloudfunctions.net/api`
+
+Append the word `/CallbackUrl` to the end of the URL, so that the end URL will be: 
+
+`https://us-central1-lnmcallback-c79b4.cloudfunctions.net/api/CallbackUrl`
+
+This will be our callback URL so that Safaricom can send us the Response to Transactions that have been initiated. Go and replace where we had written 'MY_CALLBACK_URL' with this URL.
+
+### Attaching the FirebaseMessaging service when the transaction request is successful.
```

Comment:
This and several other headings seems too long. Consider using concise headings.